core/parentchain/indirect-calls-executor/src/executor.rs

```diff
@@ -0,0 +1,436 @@
+		Ok(self.stf_enclave_signer.get_enclave_account()?)
+	}
+
+	fn sign_call_with_self(
+		&self,
+		trusted_call: &TrustedCall,
+		shard: &ShardIdentifier,
+	) -> Result<TrustedCallSigned> {
+		Ok(self.stf_enclave_signer.sign_call_with_self(trusted_call, shard)?)
+	}
+}
+
+pub(crate) fn hash_of<T: Encode>(xt: &T) -> H256 {
+	blake2_256(&xt.encode()).into()
+}
+
+#[cfg(test)]
+mod test {
+	use super::*;
+	use crate::filter_calls::ShieldFundsAndCallWorkerFilter;
+	use codec::{Decode, Encode};
+	use ita_stf::TrustedOperation;
+	use itc_parentchain_test::parentchain_block_builder::ParentchainBlockBuilder;
+	use itp_node_api::{
+		api_client::{
+			ParentchainExtrinsicParams, ParentchainExtrinsicParamsBuilder,
+			ParentchainUncheckedExtrinsic,
+		},
+		metadata::{metadata_mocks::NodeMetadataMock, provider::NodeMetadataRepository},
+	};
+	use itp_sgx_crypto::mocks::KeyRepositoryMock;
+	use itp_stf_executor::mocks::StfEnclaveSignerMock;
+	use itp_stf_primitives::types::AccountId;
+	use itp_test::mock::shielding_crypto_mock::ShieldingCryptoMock;
+	use itp_top_pool_author::mocks::AuthorApiMock;
+	use itp_types::{Block, CallWorkerFn, Request, ShardIdentifier, ShieldFundsFn};
+	use sp_core::{ed25519, Pair};
+	use sp_runtime::{MultiSignature, OpaqueExtrinsic};
+	use std::assert_matches::assert_matches;
+	use substrate_api_client::{ExtrinsicParams, GenericAddress};
+
+	type TestShieldingKeyRepo = KeyRepositoryMock<ShieldingCryptoMock>;
+	type TestStfEnclaveSigner = StfEnclaveSignerMock;
+	type TestTopPoolAuthor = AuthorApiMock<H256, H256>;
+	type TestNodeMetadataRepository = NodeMetadataRepository<NodeMetadataMock>;
+	type TestIndirectCallExecutor = IndirectCallsExecutor<
+		TestShieldingKeyRepo,
+		TestStfEnclaveSigner,
+		TestTopPoolAuthor,
+		TestNodeMetadataRepository,
+		ShieldFundsAndCallWorkerFilter,
+	>;
+
+	type Seed = [u8; 32];
+	const TEST_SEED: Seed = *b"12345678901234567890123456789012";
+
+	#[test]
+	fn indirect_call_can_be_added_to_pool_successfully() {
+		let _ = env_logger::builder().is_test(true).try_init();
+
+		let (indirect_calls_executor, top_pool_author, _) =
+			test_fixtures([0u8; 32], NodeMetadataMock::new());
+
+		let opaque_extrinsic =
+			OpaqueExtrinsic::from_bytes(call_worker_unchecked_extrinsic().encode().as_slice())
+				.unwrap();
+
+		let parentchain_block = ParentchainBlockBuilder::default()
+			.with_extrinsics(vec![opaque_extrinsic])
+			.build();
+
+		indirect_calls_executor
+			.execute_indirect_calls_in_extrinsics(&parentchain_block)
+			.unwrap();
+
+		assert_eq!(1, top_pool_author.pending_tops(shard_id()).unwrap().len());
+	}
+
+	#[test]
+	fn shielding_call_can_be_added_to_pool_successfully() {
+		let _ = env_logger::builder().is_test(true).try_init();
+
+		let mr_enclave = [33u8; 32];
+		let (indirect_calls_executor, top_pool_author, shielding_key_repo) =
+			test_fixtures(mr_enclave.clone(), NodeMetadataMock::new());
+		let shielding_key = shielding_key_repo.retrieve_key().unwrap();
```

Comment:
This is a nit that I wont harp on too much because I am guilty of the same. But even if it is valid to `unwrap()` in a test it is much better to use `expect()` and give some context. I am lazy and guilty of the same though but we should consider it in the future when writing tests.